### DISCUSSION TITLE
Suspension advice

### DISCUSSION THREAD

**Discussion by 7Bpencil** [bepu/bepuphysics2#112] [Q&A] [answered]
Hello Norbo, I tried to make a torsion bar suspension (simply attached boxes-torsion keys to the hull by Hinges and AngularServos, then attached wheels to that boxes by Hinges and AngularAxisMotors). It kinda works but I am not sure is it better than simple LinearAxisServo (like in the tank demo). What do you think? Thank you
![torsion_bars](https://user-images.githubusercontent.com/43344410/102015317-b6614880-3d7c-11eb-858a-4def6295335d.png)

**Comment by RossNordby** [comment, marked as answer]
Depends on your definition of better!

My main concern with heavily loaded multibody constraint chains would be stability and rigidity. They'll tend to need high stiffness, but that makes it harder to solve and can require a shorter timestep or substepping. Any issues with high mass ratios will be magnified by using a chain of bodies.

Judging by the simulation time breakdown, it looks like you're already using substepping, but if you aren't I'd recommend trying it out (https://github.com/bepu/bepuphysics2/blob/master/Demos/Demos/SubsteppingDemo.cs). Then you can try increasing constraint frequencies to up to half of the effective solver update rate. 60hz timestep with 8x substepping would let you do 240hz constraint frequencies safely.

Sufficiently high stiffness and substepping rates *should* let you do just about anything in a stable way, it's just a question of cost. Note that for small simulations, using only a single thread could help, since otherwise most of the simulation time could be wrangling threads. (I also have plans to reduce the cost of substepping by a lot: #104)

Could also play with different constraint configurations that get rid of the intermediate bodies, though it might require some approximations (or custom constraints) in more complex configurations.

**Comment by 7Bpencil** [comment]
I did some more tests, looks like such chain of bodies isn't stable without substepping. And I'm not sure about its usage in real apps. On the other side, such suspension moves more interestingly than LinearAxisServo. Maybe in the future, I will find the courage to create custom constraints. Thank you, for a hint!

**Comment by RossNordby** [comment]
By the way, something I should have mentioned is that a great deal of the solving difficulty in such situations arises from large mass ratios. If you were to artificially reduce the mass ratio between bodies such that the wheels and intermediate bodies were closer to the mass of the main body, it would very likely help with stability and not require as many solver iterations or substeps.

**Comment by 7Bpencil** [comment]
Silly me! With torsion key mass equals 2x wheel mass, everything is stable without substepping. But I encountered another thing: it looks like internal forces pull a structure apart - a tank slowly moves sideways on its own. Another iteration on balancing masses is needed haha. Also simulation sometimes spikes from 0.7ms to 6ms

(also it kinda funny with low-force AngularServo)
![funky_torsions](https://user-images.githubusercontent.com/43344410/103687800-ae2eae80-4fb2-11eb-9b06-61b77a2dec9f.gif)

**Comment by RossNordby** [comment]
>But I encountered another thing: it looks like internal forces pull a structure apart - a tank slowly moves sideways on its own.

This is usually a sign that some constraint configurations are fighting somewhere, or that a constraint stiffness are a little too high for the current effective solver timestep, or there is some other difficulty (like mass ratios, as you mention) that is harming convergence. If the solver can't converge to a global solution, it can introduce energy.

>Also simulation sometimes spikes from 0.7ms to 6ms

If it's happening during steady state simulation (no sleep/wake transitions, no JIT compilations, etc), that usually happens when the OS decides to switch out a worker thread, and then the entire engine sits waiting for the OS to reschedule it so the work can finish. It's made more likely by thread oversubscription. If the OS doesn't have a free thread to schedule some background work, it could displace a physics engine thread.

This kind of issue tends to be far more common on older operating systems like Windows 7. They tend to require a bit more breathing room- giving the physics a smaller subset of the available logical threads can help.

And if you only have a few tanks of that size, there's a good chance they'd run faster single threaded too. Multithreading only really starts to shine when there's thousands of constraints and bodies.

>(also it kinda funny with low-force AngularServo)

In the future, all tanks will have a theme park ride mode.

**Comment by 7Bpencil** [comment]
Yep, I eventually found sweet mass-ratio balance
![first_stable_torsions](https://user-images.githubusercontent.com/43344410/103714021-7853ef00-4fdf-11eb-92af-61eb5694ff92.PNG)


But then I realized that this janky movement is awful. Okay, maybe I will give it a try later
![janky_torsions](https://user-images.githubusercontent.com/43344410/103713630-a97fef80-4fde-11eb-9844-909dee7a918e.gif)

Thank for the help

**Comment by RossNordby** [comment]
Might be able to improve things by creating direct connections between the wheels and the main body. An AngularAxisMotor for the drive torque, AngularHinge to keep the wheels aligned, and maybe a LinearAxisServo to constrain the center of the wheels to stay on on the proper plane. The suspension would then only be responsible for movement along the 2 remaining degrees of freedom on that plane. Might be possible to then simplify the connections to the suspension body.

**Comment by 7Bpencil** [comment]
Okay, I have eventually set AngularServo max speed to 8 instead of float.MaxValue and it minimized those janky movements. But I found that tank doesn't go straight:
![not_straight](https://user-images.githubusercontent.com/43344410/104859778-619b8980-5949-11eb-99d2-38a7967424b8.gif)
I give up for now. Let's hope that no one will notice.
